google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/DdlBatch.java

```diff
@@ -168,6 +169,17 @@ public Timestamp getCommitTimestampOrNull() {
     return null;
   }
 
+  @Override
+  public CommitResponse getCommitResponse() {
+    throw SpannerExceptionFactory.newSpannerException(
+        ErrorCode.FAILED_PRECONDITION, "There is no commit response available for DDL batches.");
```

Comment:
This is a code smell suggests that perhaps getCommitResponse does not belong in the superclass/interface in the first place since it doesn't apply to all subclasses.

Reply:
This is a package private class (extending from a package private interface), so it is not something that client applications will deal with. Client applications only interact with the [Connection](https://github.com/googleapis/java-spanner/blob/1b077e457d40dd1f64c20a9f7d897313f78ea4e9/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/Connection.java#L141) interface. (And client applications in this case is only intended to be frameworks /drivers that implement a declarative connection-type interface, such as JDBC.)
These declarative style APIs often mean that certain methods are only applicable in certain states, and this is what is also reflected in these classes. The method that client applications will interact with is [Connection#getCommitResponse()](https://github.com/googleapis/java-spanner/blob/1b077e457d40dd1f64c20a9f7d897313f78ea4e9/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/Connection.java#L654). This method will delegate the actual response to the at that moment active `UnitOfWork` on the `Connection`. The different `UnitOfWork` implementations will respond based on what they support, which may also be that they do not support it.

Reply:
Have you considered returning an empty object or null instead?

If clients really won't see this, fine. However it's not enough that this class is non-public. They could still invoke this method if they get an instance of it, even while only knowing its supertype.

Reply:
Returning null or an empty object here would only move the logic to the `ConnectionImpl` class, which should throw an exception if it is not in a valid state to return `CommitStats` at that moment (for example because a DDL batch is active). This logic is therefore delegated to the concrete implementations of `UnitOfWork`.

This class and its super types are all package private and or not accessible from outside the library.